pkg/apis/discovery/validation/validation.go

```diff
@@ -53,29 +52,44 @@ var (
 var ValidateEndpointSliceName = apimachineryvalidation.NameIsDNSSubdomain
 
 // ValidateEndpointSlice validates an EndpointSlice.
-func ValidateEndpointSlice(endpointSlice *discovery.EndpointSlice, validAddressTypes sets.String) field.ErrorList {
+func ValidateEndpointSlice(endpointSlice *discovery.EndpointSlice, allowNodeName bool) field.ErrorList {
 	allErrs := apivalidation.ValidateObjectMeta(&endpointSlice.ObjectMeta, true, ValidateEndpointSliceName, field.NewPath("metadata"))
-	allErrs = append(allErrs, validateAddressType(endpointSlice.AddressType, validAddressTypes)...)
-	allErrs = append(allErrs, validateEndpoints(endpointSlice.Endpoints, endpointSlice.AddressType, field.NewPath("endpoints"))...)
+	allErrs = append(allErrs, validateAddressType(endpointSlice.AddressType)...)
+	allErrs = append(allErrs, validateEndpoints(endpointSlice.Endpoints, endpointSlice.AddressType, allowNodeName, field.NewPath("endpoints"))...)
 	allErrs = append(allErrs, validatePorts(endpointSlice.Ports, field.NewPath("ports"))...)
 
 	return allErrs
 }
 
 // ValidateEndpointSliceCreate validates an EndpointSlice when it is created.
 func ValidateEndpointSliceCreate(endpointSlice *discovery.EndpointSlice) field.ErrorList {
-	return ValidateEndpointSlice(endpointSlice, supportedAddressTypes)
+	// allow NodeName value if the feature gate is set.
+	allowNodeName := utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
+
+	return ValidateEndpointSlice(endpointSlice, allowNodeName)
 }
 
 // ValidateEndpointSliceUpdate validates an EndpointSlice when it is updated.
 func ValidateEndpointSliceUpdate(newEndpointSlice, oldEndpointSlice *discovery.EndpointSlice) field.ErrorList {
-	allErrs := ValidateEndpointSlice(newEndpointSlice, supportedAddressTypes.Union(deprecatedAddressTypes))
+	// allow NodeName value if the feature gate is set.
+	allowNodeName := utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
+
+	if !allowNodeName {
```

Comment:
this means that if we have the feature gate disabled for node name, but one of the old slices has the field set, we allow node name?

Reply:
ic, there is also a test, I always find this part confusing so take it with a grain of salt, but I thought that the recommended way is to clear the field in strategy and only validate it here is is not nil 
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version

Reply:
> I thought that the recommended way is to clear the field in strategy and only validate it here is is not nil

Agreed, I thought we need to drop disabled fields in strategy, the validation logic here seems fine though. Left a related comment here https://github.com/kubernetes/kubernetes/pull/96440#discussion_r521423193

Reply:
Yep, I believe this is intended to help with a scenario where a newer client hits an older API server. I've seen this similar pattern used in several places, including my own [use of it with AppProtocol](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/apis/core/validation/validation.go#L4180-L4187).

Reply:
My local copy of this page missed the follow up comments here before I commented - will update strategy here as well, thanks for the references!

Reply:
Strategy has been updated.